arangod/Aql/Executor/EnumerateNearVectorExecutor.cpp

```diff
@@ -41,132 +49,197 @@ using Stats = NoStats;
+                    vectorComponentsCount));
   }
 }
 
-void EnumerateNearVectorsExecutor::searchResults(
-    std::vector<float>& inputRowsJoined) {
+void EnumerateNearVectorsExecutor::searchResults() {
   auto* vectorIndex = dynamic_cast<RocksDBVectorIndex*>(_infos.index.get());
   TRI_ASSERT(vectorIndex != nullptr);
 
-  std::vector<float> convertedInputRows(_inputRows.size());
   RocksDBMethods* mthds =
       RocksDBTransactionState::toMethods(&_trx, _collection->id());
 
   std::tie(_labels, _distances) = vectorIndex->readBatch(
-      inputRowsJoined, mthds, &_trx, _collection->getCollection(),
-      _inputRows.size(), _infos.topK);
-
-  _initialized = true;
-  _currentProcessedResultCount = 0;
+      _inputRowConverted, mthds, &_trx, _collection->getCollection(), 1,
+      _infos.getNumberOfResults());
+  LOG_INTERNAL << "Results: " << _labels << " and distances: " << _distances;
 }
 
 void EnumerateNearVectorsExecutor::fillOutput(OutputAqlItemRow& output) {
   auto const docOutId = _infos.outDocumentIdReg;
   auto const distOutId = _infos.outDistancesReg;
 
-  auto inputRowIterator = _inputRows.begin();
   while (!output.isFull() &&
-         _currentProcessedResultCount < _inputRows.size() * _infos.topK) {
+         _currentProcessedResultCount < _infos.getNumberOfResults()) {
     // there are no results anymore for this input, so we can skip to next input
     // row
     if (_labels[_currentProcessedResultCount] == -1) {
-      _currentProcessedResultCount =
-          (_currentProcessedResultCount / _infos.topK + 1) * _infos.topK;
-      ++inputRowIterator;
-      continue;
+      _currentProcessedResultCount = _infos.getNumberOfResults();
+      break;
     }
     output.moveValueInto(
-        docOutId, *inputRowIterator,
+        docOutId, _inputRow,
         AqlValueHintUInt(_labels[_currentProcessedResultCount]));
     output.moveValueInto(
-        distOutId, *inputRowIterator,
+        distOutId, _inputRow,
         AqlValueHintDouble(_distances[_currentProcessedResultCount]));
     output.advanceRow();
 
     ++_currentProcessedResultCount;
-    if (_currentProcessedResultCount % _infos.topK == 0) {
-      ++inputRowIterator;
-    }
+  }
+
+  if (_currentProcessedResultCount == _infos.getNumberOfResults()) {
+    _resultsAreProcessed = true;
   }
 }
 
 std::tuple<ExecutorState, NoStats, AqlCall>
 EnumerateNearVectorsExecutor::produceRows(AqlItemBlockInputRange& inputRange,
                                           OutputAqlItemRow& output) {
-  NoStats stats;
+  LOG_INTERNAL << "hasDataRow: " << inputRange.hasDataRow();
+  while (!output.isFull()) {
+    if (!_initialized) {
+      if (!_inputRow.isInitialized()) {
+        fillInput(inputRange);
+      }
+
+      searchResults();
+    }
+    fillOutput(output);
+    if (_currentProcessedResultCount == _infos.getNumberOfResults()) {
+      inputRange.advanceDataRow();
+      _initialized = false;
+      break;
+    }
+  }
```

Comment:
For one input we would have `_infos.getNumberOfResults()` number of results to write to output, meaning we don't read input again until we have reached `_infos.getNumberOfResults()`. We fill output from the results in `searchResults` function.
So I would say no, since we are interested in `inputData` every `_infos.getNumberOfResults()` number of iterations

Reply:
But it's not a given that there are even `_infos.getNumberOfResults()` many documents in the index/collection, right?

Reply:
Yes, that is true, but faiss will always return that many documents ids. returning -1 for those which could not have been fetched. And if we detect that we would skip all of invalid documents